Add documentation for `CodeGenerator`s. (#299)
square/anvil: README.md

exclusion rule does what it implies. In this specific example `DaggerModule` wishes to be
contributed to this scope, but it has been excluded for this component and thus is not added.

> ## Dagger Factory generation

Reviewer: ```suggestion
## Dagger Factory Generation
```

Either cap them all or only capitalize Dagger

of Java code, which allows Gradle to skip the Java compilation task. The result is faster
builds.

```groovy
anvil {
  generateDaggerFactories = true // default is false
}
```

In our codebase we measured that modules using Dagger build 65% faster with this new Anvil feature
compared to using the Dagger annotation processor:

| Stub generation | Kapt | Javac | Kotlinc | Sum
:--- | ---: | ---: | ---: | ---: | ---:
Dagger | 12.976 | 40.377 | 8.571 | 10.241 | 72.165
Anvil | 0 | 0 | 6.965 | 17.748 | 24.713

For full builds of applications we measured savings of 16% on average.

![Benchmark Dagger Factories](images/benchmark_dagger_factories.png?raw=true "Benchmark Dagger Factories")

This feature can only be enabled in Gradle modules that don't compile any Dagger component. Since
Anvil only processes Kotlin code, you shouldn't enable it in modules with mixed Kotlin / Java
sources either.

When you enable this feature, don't forget to remove the Dagger annotation processor. You should
keep all other dependencies.

## Extending Anvil

Every codebase has its own dependency injection patterns where certain code structures need to be

> repeated over and over again. Here Anvil comes to the rescue, and you can extend the compiler 

Reviewer: ```suggestion
repeated over and over again. Here Anvil comes to the rescue and you can extend the compiler 
```

> plugin with your own code generator. For steps please take a look at the 

Reviewer: ```suggestion
plugin with your own `CodeGenerator`. For usage please take a look at the 
```

![Benchmark](images/benchmark.png?raw=true "Benchmark")

> On top of that Anvil provides actual build time improvements by replacing the Dagger annotation

Reviewer: ```suggestion
On top of that, Anvil provides actual build time improvements by replacing the Dagger annotation
```
